### PR TITLE
Add Apigee X VPN Hub and Spoke Design TF scripts

### PR DESCRIPTION
**Pull request by judyzhaoxinwu**
Add scripts for how to create an Apigee Organization, connect the Apigee Runtime to GCP VPC, connect Apigee VPC with multiple backends through Cloud VPN, and set up DNS peering.

### CODE REVIEW

File: examples/apigee-x-vpn-hub-and-spoke-design/.terraform.lock.hcl [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# This file is maintained automatically by "terraform init".
+# Manual edits may be lost in future updates.
+
+provider "registry.terraform.io/hashicorp/google" {
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/.terraform.lock.hcl:4]: lockfile should not be checked in

File: examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project.tf [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+# Your use of it is subject to your agreement with Google.
+#
+# Sample Terraform script to set up an Apigee X instance 
+##########################################################
+
+##########################################################
+### Create a Google Cloud Project for Apigee X runtime
+##########################################################
+
+resource "google_project" "apigee_x_project" {
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project.tf:14]: You can use https://github.com/terraform-google-modules/terraform-google-project-factory which has some of the features (like random suffix), service enablement etc built in

File: examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project.tf [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  provider = google
+  project  = google_project.apigee_x_project.project_id
+  service  = "cloudkms.googleapis.com"
+}
+
+resource "google_project_service" "dns" {
+  provider = google
+  project  = google_project.apigee_x_project.project_id
+  service  = "dns.googleapis.com"
+}
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project.tf:58]: nit: newlines here and throughout

File: examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project_subnets.tf [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+# representation for any use or purpose.
+# Your use of it is subject to your agreement with Google.
+#
+# Sample Terraform script to set up an Apigee X instance 
+##########################################################
+
+#######################################################################
+### Create 2 subnets, firewall rules and test instances in Apigee X project
+#######################################################################
+resource "google_compute_subnetwork" "apigee_x_project_subnet1" {
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project_subnets.tf:13]: You can use https://github.com/terraform-google-modules/terraform-google-network

File: examples/apigee-x-vpn-hub-and-spoke-design/backend_project_a_vpn_subnet1.tf [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+  project                   = google_project.backend_project_a.project_id
+  name                      = "backend-project-a-router1-peer2"
+  router                    = google_compute_router.backend_project_a_router1.name
+  region                    = "${var.subnet_1}"
+  peer_ip_address           = "169.254.1.2"
+  peer_asn                  = 64514
+  advertised_route_priority = 100
+  interface                 = google_compute_router_interface.backend_project_a_router1_interface2.name
+}
+
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/backend_project_a_vpn_subnet1.tf:154]: empty lines below this

File: examples/apigee-x-vpn-hub-and-spoke-design/terraform.tfstate [hunk 6]
```diff
@@ -0,0 +1,1834 @@
+{
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/terraform.tfstate:1]: state should not be checked in

File: tools/table_validator/README.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
+# Teradata to BigQuery Table Validator
```
Comment (bharathkkb) [inline tools/table_validator/README.md:1]: These look like unrelated changes

File: examples/apigee-x-vpn-hub-and-spoke-design/variables.tf [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  description = "The CIDR mask for creating Apigee instance"
+  type        = number
+}
+ 
+variable "subnet_1" {
+  description = "The region for a subnet"
+  type        = string
+}
+
+variable "subnet_2" {
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/variables.tf:37]: If this is an example you can provide defaults for these rather than requiring inputs

File: examples/apigee-x-vpn-hub-and-spoke-design/versions.tf [hunk 9]
```diff
@@ -0,0 +1,17 @@
+terraform {
+  required_version = ">= 0.13"
+  required_providers {
+    google = {
+      source  = "hashicorp/google"
+      version = "4.0.0"
+    }
+    # google-beta = {
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/versions.tf:8]: nit: commented code

File: examples/apigee-x-vpn-hub-and-spoke-design/versions.tf [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    #   source  = "hashicorp/google"
+    #   version = "3.87.0"
+    # }
+  }
+}
+
+provider "google" {
+  
+}
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/versions.tf:17]: empty provider blocks are not necessary

File: examples/apigee-x-vpn-hub-and-spoke-design/dns_peering.tf [hunk 11]
```diff
@@ -0,0 +1,39 @@
+resource "google_dns_managed_zone" "apigee_x_project_forwarding_zone" {
+  project     = google_project.apigee_x_project.project_id
+  name        = "apigee-x-project-forwarding-zone"
+  dns_name    = "${var.private_zone_domain}"
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/dns_peering.tf:4]: Interpolation is not necessary here and throughtout
```suggestion
  dns_name    = var.private_zone_domain
```

File: examples/apigee-x-vpn-hub-and-spoke-design/dns_peering.tf [hunk 12]
```diff
@@ -0,0 +1,39 @@
+resource "google_dns_managed_zone" "apigee_x_project_forwarding_zone" {
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/dns_peering.tf:1]: missing headers

File: examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project_vpc.tf [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+  project    = google_project.apigee_x_project.project_id
+  name       = "apigee-x-keyring"
+  location   = "us-central1"
+
+  lifecycle {
+    prevent_destroy = true
+  }
+}
+
+resource "google_kms_crypto_key" "apigee_x_key" {
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project_vpc.tf:69]: KMS and IAM on keys can be managed via https://github.com/terraform-google-modules/terraform-google-kms

File: examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project_subnets.tf [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+  allow {
+    protocol = "icmp"
+  }
+
+  source_ranges = ["0.0.0.0/0"]
+
+  target_tags   = ["apigee-x-vpc-allow-icmp"]
+}
+
+resource "google_compute_instance" "instance-01" {
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/apigee_x_project_subnets.tf:64]: name for file is misleading as it has a subnets prefix but you are also creating compute instances

File: examples/apigee-x-vpn-hub-and-spoke-design/README.md [hunk 15]
```diff
@@ -0,0 +1,40 @@
+# Apigee X VPN Hub and Spoke Design
+The purspose is to showcase that multiple backends can use Cloud VPN to connect to the Apigee X VPC and then connect to the Apigee Runtime. 
+
+## Setup
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/README.md:4]: We should add instructions on how to run the example

**Comment by juliocc** [comment]
Is there anything significant in this PR that is not available in the [Apigee samples repo](https://github.com/apigee/terraform-modules)?

File: examples/apigee-x-vpn-hub-and-spoke-design/main.tf [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+# representation for any use or purpose.
+# Your use of it is subject to your agreement with Google.
+#
+# Sample Terraform script to set up an Apigee X instance 
+##########################################################
+
+##########################################################
+### Create a Google Cloud Project for the demo
+##########################################################
+data "google_client_config" "current" {}
```
Comment (bharathkkb) [inline examples/apigee-x-vpn-hub-and-spoke-design/main.tf:13]: Is this used anywhere?

**Review by bharathkkb — COMMENTED** [review]
Some initial comments, I'll defer to the owners for other suggestions

File: examples/apigee-x-vpn-hub-and-spoke-design/.gitignore [hunk 17]
```diff
@@ -0,0 +1,4 @@
+*.tfstate
+*.tfstate.backup
+.terraform.lock.hcl
+.terraform/
```
Comment (boredabdel) [inline examples/apigee-x-vpn-hub-and-spoke-design/.gitignore:4]: Please replace the last line with **/.terraform/*
This way providers will not be included in the commit. They are quite big files

This should exclude the lock file as well.

File: examples/apigee-x-vpn-hub-and-spoke-design/README.md (hunk 15 above)
Comment (boredabdel) [reply examples/apigee-x-vpn-hub-and-spoke-design/README.md:4]: Yes please.

File: examples/apigee-x-vpn-hub-and-spoke-design/README.md [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+`apigee_x_project_subnets.tf` creates two subnets (us-east1 and us-west2) within the Apigee VPC and one VMs in each zone for testing the connectivity. Please note that Apigee Runtime requires ssh and icmp connections. 
+
+### VPN Gateways
+In order to connect to other VPCs from the Apigee VPC, HA Tunnels are needed in this case. `apigee_x_project_router_vpn.tf` sets up a Cloud Router and VPN gateway in each zone for the Apigee X VPC. `backend_project_a.tf` creates a GCP project and a VPC for the Backend a. `backend_project_a_vpn_subnet1.tf` and `backend_project_a_vpn_subnet2.tf` will create th VPN gateway interfaces, tunnels and router peering.
+
+_If the connections are established, you will be able to ping the VM in the same zone from the other VPC._
+
+
+### DNS Peering
+WIP
```
Comment (boredabdel) [inline examples/apigee-x-vpn-hub-and-spoke-design/README.md:37]: Are you planning to add these as part of this PR or later on ?

File: examples/apigee-x-vpn-hub-and-spoke-design/backend_project_a_vpn_subnet2.tf [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+  project                   = google_project.backend_project_a.project_id
+  name                      = "backend-project-a-router2-peer2"
+  router                    = google_compute_router.backend_project_a_router2.name
+  region                    = "${var.subnet_2}"
+  peer_ip_address           = "169.254.1.2"
+  peer_asn                  = "${var.apigee_x_project_router2_asn}"
+  advertised_route_priority = 100
+  interface                 = google_compute_router_interface.backend_project_a_router2_interface2.name
+}
+
```
Comment (boredabdel) [inline examples/apigee-x-vpn-hub-and-spoke-design/backend_project_a_vpn_subnet2.tf:154]: Please remove empty lines below and leave 1 line

File: examples/apigee-x-vpn-hub-and-spoke-design/.gitignore (hunk 17 above)
Comment (boredabdel) [reply examples/apigee-x-vpn-hub-and-spoke-design/.gitignore:4]: Also add these two lines

*.tfstate
*.tfstate.*

To avoid committing tfstate files.

**Review by boredabdel — COMMENTED** [review]
Please have a look at the comments below.

The PR also seems to included changes from unrelated folders. Please have a look at those.

**Comment by boredabdel** [comment]
As discussed offline this will be committed to a different repo. Closing this PR
